File: node/import-csv-to-postgres.js
const fs = require('fs');
const argparse = require('argparse');
const { Client } = require('pg');

function getTableColumns(client, table) {
  return client.query(`SELECT * FROM ${table} LIMIT 0`)
    .then(result => result.fields.map(field => field.name));
}

async function insertCSVData(csvFilename, dbParams, table) {
  const client = new Client(dbParams);
  
  try {
    await client.connect();
    console.log("Connected to the database successfully!");

    const tableColumns = await getTableColumns(client, table);
    const csvData = await fs.promises.readFile(csvFilename, 'utf8');
    const csvLines = csvData.split('\n');
    const csvHeader = csvLines[0].split(',');

    function cleanColumnName(columnName) {
      return columnName.replace(/[^a-zA-Z0-9]/g, '');
    }

    const cleanedCsvHeader = csvHeader.map(cleanColumnName);
    const cleanedTableColumns = tableColumns.map(cleanColumnName);

    if (JSON.stringify(cleanedCsvHeader) === JSON.stringify(cleanedTableColumns)) {
      for (let i = 1; i < csvLines.length; i++) {
        const row = csvLines[i].split(',');
        const query = `INSERT INTO ${table} VALUES (${Array(row.length).fill().map((_, idx) => `$${idx + 1}`).join(', ')})`;
        await client.query(query, row);
      }
      console.log("Data inserted successfully!");
    } else {
      console.log("CSV header does not match table columns:");
      console.log("CSV Header:", cleanedCsvHeader);
      console.log("Table Columns:", cleanedTableColumns);
    }

  } catch (error) {
    console.error("Error:", error);
  } finally {
    client.end();
  }
}

const parser = new argparse.ArgumentParser({
  description: 'Insert CSV data into PostgreSQL'
});
parser.addArgument(['--csv'], { required: true, help: 'CSV filename' });
parser.addArgument(['--dbname'], { required: true, help: 'Database name' });
parser.addArgument(['--user'], { required: true, help: 'Username' });
parser.addArgument(['--password'], { required: true, help: 'Password' });
parser.addArgument(['--host'], { required: true, help: 'Host' });
parser.addArgument(['--port'], { required: true, help: 'Port' });
parser.addArgument(['--table'], { required: true, help: 'Table name' });
const args = parser.parseArgs();

const dbParams = {
  database: args.dbname,
  user: args.user,
  password: args.password,
  host: args.host,
  port: args.port
};

insertCSVData(args.csv, dbParams, args.table);

File: python/import-csv-to-postgres.py
import csv
import argparse
import psycopg2
import re

def get_table_columns(cur, table_name):
    cur.execute(f"SELECT * FROM {table_name} LIMIT 0")
    return [desc[0] for desc in cur.description]

def clean_column_name(column_name):
    return re.sub(r'[^a-zA-Z0-9]', '', column_name)

def insert_csv_data(csv_filename, db_params, table_name):
    try:
        conn = psycopg2.connect(**db_params)
        print("Connected to the database successfully!")
        cur = conn.cursor()

        table_columns = get_table_columns(cur, table_name)
        
        with open(csv_filename, 'r') as csv_file:
            csv_reader = csv.reader(csv_file)
            csv_header = next(csv_reader)

            if all(clean_column_name(col) for col in csv_header) == all(clean_column_name(col) for col in table_columns):                
                for row in csv_reader:
                    query = f"INSERT INTO {table_name} VALUES ({', '.join(['%s']*len(row))})"
                    cur.execute(query, row)

                conn.commit()
                print("Data inserted successfully!")

            else:
                print("CSV header does not match table columns:")
                print("CSV Header:", csv_header)
                print("Table Columns:", table_columns)

        cur.close()
        conn.close()

    except (Exception, psycopg2.Error) as error:
        print("Error:", error)

    finally:
        if conn:
            conn.close()

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Insert CSV data into PostgreSQL')
    parser.add_argument('--csv', required=True, help='CSV filename')
    parser.add_argument('--dbname', required=True, help='Database name')
    parser.add_argument('--user', required=True, help='Username')
    parser.add_argument('--password', required=True, help='Password')
    parser.add_argument('--host', required=True, help='Host')
    parser.add_argument('--port', required=True, help='Port')
    parser.add_argument('--table', required=True, help='Table name')
    args = parser.parse_args()

    db_params = {
        'dbname': args.dbname,
        'user': args.user,
        'password': args.password,
        'host': args.host,
        'port': args.port
    }

    insert_csv_data(args.csv, db_params, args.table)
